_posts/2015-11-09-Generators.md

```diff
@@ -0,0 +1,250 @@
+  } else {
+    if (typeof value === "undefined") {
+      throw new Error("`value` must be a number.");
+    }
+    take10.push(value); // no error
+  }
+}
+```
+
+There is an [open issue](https://github.com/facebook/flow/issues/577) which would make the dynamic type test above unnecessary, by using the `done` value as a sentinel to refine a tagged union. That is, when `done` is `true`, Flow would know that `value` is always of type `Yield` and otherwise of type `Return | void`.
+
+Even without the dynamic type test, this code is quite verbose and it's hard to see the intent. Because generators are also iterable, we can also use `for...of` loops:
+
+```javascript
+const take10: number[] = [];
+let i = 0;
+for (let nat of nats()) {
+  if (i === 10) break;
+  take10.push(nat);
+  i++;
+}
+```
+
+That's much better. The `for...of` looping construct ignores completion values, so Flow understands that `nat` will always be `number`. Let's generalize this pattern further using generator functions:
+
+```javascript
+function *take<T>(n: number, xs: Iterable<T>): Iterable<T> {
+  if (n <= 0) return;
+  let i = 0;
+  for (let x of xs) {
+    yield x;
+    if (++i === n) return;
+  }
+}
+
+for (let n of take(10, nats())) {
+  console.log(n); // 0, 1, 2, 3, 4, 5, 6, 7, 8, 9
+}
+```
+
+Note that we explicitly annotated the parameters and return type of the `take` generator. This is necessary to ensure Flow understands the fully generic type.
+
+Another important feature of generators is the ability to pass values into the generator from the consumer. Let's consider a generator `scan`, which reduces values passed into the generator using a provided function. Our `scan` is similar to `Array.prototype.reduce`, but it returns each intermediate value and the values are provided imperatively via `next`.
+
+As a first pass, we might write this:
+
+```javascript
+function *scan<T,U>(init: U, f: (acc: U, x: T) => U): Generator<U,void,T> {
+  let acc = init;
+  while (true) {
+    const next = yield acc;
+    acc = f(acc, next);
+  }
+}
+```
+
+We can use this definition to implement an imperative sum procedure:
+
+```javascript
+let sum = scan(0, (a,b) => a + b);
+console.log(sum.next());  // { done: false, value: 0 }
+console.log(sum.next(1)); // { done: false, value: 1 }
+console.log(sum.next(2)); // { done: false, value: 3 }
+console.log(sum.next(3)); // { done: false, value: 6 }
+```
+
+However, when we try to check the above definition of `scan`, Flow complains:
+
+```
+test.js:7
+  7:     acc = f(acc, next);
+               ^^^^^^^^^^^^ function call
+  7:     acc = f(acc, next);
+                      ^^^^ undefined. This type is incompatible with
+  3: function *scan<T,U>(init: U, f: (acc: U, x: T) => U): Generator<U,void,T> {
+     ^ some incompatible instantiation of T
+```
+
+Flow is complaining that our value, `next`, may be `void` instead of the expected `T`, which is `number` in the `sum` example. This is unfortunate, but necessary for type safety.
```

Comment:
Nit: No need to frame apologetic, can probably drop the "This is infortunate" and just clarify why